File: changmen/gamebet_backend/shared/storage_paths.js
"use strict";

const fs = require("fs");
const path = require("path");

const BACKEND_ROOT = process.env.GAMEBET_BACKEND_ROOT
  || path.join(__dirname, "..");

const CHANGMEN_ROOT = process.env.GAMEBET_CHANGMEN_ROOT
  || path.dirname(BACKEND_ROOT);

const STORAGE_DIR = process.env.GAMEBET_STORAGE_DIR
  || path.join(BACKEND_ROOT, "storage");

const DB_DIR = process.env.GAMEBET_DB_DIR
  || path.join(CHANGMEN_ROOT, "gamebetdb");
const LEGACY_DIR = path.join(STORAGE_DIR, "legacy");
const LEGACY_ESPORT_DIR = path.join(LEGACY_DIR, "esport");

const OLD_DATA_DIR = path.join(BACKEND_ROOT, "data");
const OLD_DB_PATH = path.join(OLD_DATA_DIR, "gamebet.db");
const STORAGE_DB_PATH = path.join(STORAGE_DIR, "db", "gamebet.db");
const OLD_ESPORT_DATA_DIR = path.join(OLD_DATA_DIR, "esport");

const DB_PATH = process.env.GAMEBET_DB_PATH
  || path.join(DB_DIR, "gamebet.db");

const ESPORT_DATA_DIR = process.env.ESPORT_DATA_DIR
  || LEGACY_ESPORT_DIR;

function copyDirContentsIfMissing(fromDir, toDir) {
  if (!fs.existsSync(fromDir)) return;
  fs.mkdirSync(toDir, { recursive: true });
  for (const name of fs.readdirSync(fromDir)) {
    const fromPath = path.join(fromDir, name);
    const toPath = path.join(toDir, name);
    if (fs.existsSync(toPath)) continue;
    const stat = fs.statSync(fromPath);
    if (stat.isDirectory()) fs.cpSync(fromPath, toPath, { recursive: true });
    else fs.copyFileSync(fromPath, toPath);
  }
}

function copyFileIfMissing(fromFile, toFile) {
  if (!fs.existsSync(fromFile) || fs.existsSync(toFile)) return;
  fs.mkdirSync(path.dirname(toFile), { recursive: true });
  fs.copyFileSync(fromFile, toFile);
}

function copyDbIfMissing(fromDbPath) {
  if (!fs.existsSync(fromDbPath) || fs.existsSync(DB_PATH)) return;
  copyFileIfMissing(fromDbPath, DB_PATH);
  copyFileIfMissing(`${fromDbPath}-wal`, `${DB_PATH}-wal`);
  copyFileIfMissing(`${fromDbPath}-shm`, `${DB_PATH}-shm`);
}

function ensureStoragePaths() {
  fs.mkdirSync(path.dirname(DB_PATH), { recursive: true });
  fs.mkdirSync(ESPORT_DATA_DIR, { recursive: true });

  if (!process.env.ESPORT_DATA_DIR) {
    copyDirContentsIfMissing(OLD_ESPORT_DATA_DIR, ESPORT_DATA_DIR);
  }

  if (!process.env.GAMEBET_DB_PATH) {
    copyDbIfMissing(STORAGE_DB_PATH);
    copyDbIfMissing(OLD_DB_PATH);
  }
}

ensureStoragePaths();

module.exports = {
  BACKEND_ROOT,
  CHANGMEN_ROOT,
  STORAGE_DIR,
  DB_DIR,
  LEGACY_DIR,
  DB_PATH,
  ESPORT_DATA_DIR,
  OLD_DATA_DIR,
  OLD_DB_PATH,
  STORAGE_DB_PATH,
  OLD_ESPORT_DATA_DIR,
  ensureStoragePaths,
};

File: gamebet_backend_python/shared/storage_paths.py
import os
import shutil
from pathlib import Path

BACKEND_ROOT = Path(os.environ.get("GAMEBET_BACKEND_ROOT") or Path(__file__).parent.parent)
CHANGMEN_ROOT = Path(os.environ.get("GAMEBET_CHANGMEN_ROOT") or BACKEND_ROOT.parent)

STORAGE_DIR = Path(os.environ.get("GAMEBET_STORAGE_DIR") or BACKEND_ROOT / "storage")
DB_DIR = Path(os.environ.get("GAMEBET_DB_DIR") or CHANGMEN_ROOT / "gamebetdb")

LEGACY_DIR = STORAGE_DIR / "legacy"
LEGACY_ESPORT_DIR = LEGACY_DIR / "esport"

OLD_DATA_DIR = BACKEND_ROOT / "data"
OLD_DB_PATH = OLD_DATA_DIR / "gamebet.db"
STORAGE_DB_PATH = STORAGE_DIR / "db" / "gamebet.db"
OLD_ESPORT_DATA_DIR = OLD_DATA_DIR / "esport"

DB_PATH = Path(os.environ.get("GAMEBET_DB_PATH") or DB_DIR / "gamebet.db")
ESPORT_DATA_DIR = Path(os.environ.get("ESPORT_DATA_DIR") or LEGACY_ESPORT_DIR)


def _copy_file_if_missing(src: Path, dst: Path):
    if not src.exists() or dst.exists():
        return
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(src, dst)


def _copy_dir_contents_if_missing(src: Path, dst: Path):
    if not src.exists():
        return
    dst.mkdir(parents=True, exist_ok=True)
    for item in src.iterdir():
        target = dst / item.name
        if target.exists():
            continue
        if item.is_dir():
            shutil.copytree(item, target)
        else:
            shutil.copy2(item, target)


def _copy_db_if_missing(from_db: Path):
    if not from_db.exists() or DB_PATH.exists():
        return
    _copy_file_if_missing(from_db, DB_PATH)
    _copy_file_if_missing(Path(str(from_db) + "-wal"), Path(str(DB_PATH) + "-wal"))
    _copy_file_if_missing(Path(str(from_db) + "-shm"), Path(str(DB_PATH) + "-shm"))


def ensure_storage_paths():
    DB_PATH.parent.mkdir(parents=True, exist_ok=True)
    ESPORT_DATA_DIR.mkdir(parents=True, exist_ok=True)

    if not os.environ.get("ESPORT_DATA_DIR"):
        _copy_dir_contents_if_missing(OLD_ESPORT_DATA_DIR, ESPORT_DATA_DIR)

    if not os.environ.get("GAMEBET_DB_PATH"):
        _copy_db_if_missing(STORAGE_DB_PATH)
        _copy_db_if_missing(OLD_DB_PATH)


ensure_storage_paths()
